### PR TITLE
HIVE-25256: Support ALTER TABLE CHANGE COLUMN for Iceberg

### PR DESCRIPTION
**Pull request by marton-bod**
### What changes were proposed in this pull request?
Introduce support for `ALTER TABLE tbl CHANGE COLUMN`. This does not yet support changing fields of complex types.

### Why are the changes needed?
Allows us to rename a column, change its type (only float->double, int->bigint, decimal changes), change its comment or change its order.

### Does this PR introduce _any_ user-facing change?
yes, new query type for iceberg

### How was this patch tested?
Unit tests

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 1]
```diff
@@ -505,19 +512,82 @@ private void handleReplaceColumns(org.apache.hadoop.hive.metastore.api.Table hms
...
+          schemaDifference.getMissingFromSecond().get(0).getName(),
+          schemaDifference.getMissingFromFirst().get(0).getName());
+    }
+    Pair<String, Optional<String>> outOfOrder = HiveSchemaUtil.getFirstOutOfOrderColumn(hmsCols, icebergCols,
+        renameMapping);
+
+    if (!schemaDifference.isEmpty() || outOfOrder != null) {
+      updateSchema = icebergTable.updateSchema();
+    } else {
+      // we should get here if the user restated the exactly the existing column in the CHANGE COLUMN command
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:541]: Please fix the comment, I do not get it 😄

Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:541]: Or fix me 😄

Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:541]: If the comment is not clear to you, it needs to be fixed :) Will do it

File: hbase-handler/src/test/results/negative/hbase_ddl.q.out [hunk 2]
```diff
@@ -26,4 +26,4 @@ key                 	int                 	It is a column key
 value               	string              	It is the column string value
 	 	 
 #### A masked pattern was here ####
-FAILED: SemanticException [Error 10134]: ALTER TABLE can only be used for [ADDPROPS, DROPPROPS, ADDCOLS, REPLACE_COLUMNS, SETPARTITIONSPEC] to a non-native table  hbase_table_1
+FAILED: SemanticException [Error 10134]: ALTER TABLE can only be used for [ADDPROPS, DROPPROPS, ADDCOLS] to a non-native table  hbase_table_1
```
Comment (pvary) [inline hbase-handler/src/test/results/negative/hbase_ddl.q.out:29]: What happens with HBase tables if we try replacing columns and setting partition spec?

Comment (marton-bod) [reply hbase-handler/src/test/results/negative/hbase_ddl.q.out:29]: They get this SemanticException:
```
ALTER TABLE can only be used for [ADDPROPS, DROPPROPS, ADDCOLS] to a non-native table  hbase_table_1
```
HBase (and all other storage handlers except for Iceberg at the moment), should get this exception for alter commands other than SET/UNSET TBLPROPERTIES and ADD COLUMNS.

Comment (marton-bod) [reply hbase-handler/src/test/results/negative/hbase_ddl.q.out:29]: Previously, when we were working on adding new alter commands for Iceberg, we kept adding these new operation types (rename columns, etc.) to the allowed list. However, there was only one global allowed list for all storage handler types. Now, the allowed list has been moved into the storage handler, so I've reverted the global list to its original form (before all our Iceberg changes started flowing in)

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 1 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: Updated the comment, let me know if this clarifies it.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 3]
```diff
@@ -505,19 +512,83 @@ private void handleReplaceColumns(org.apache.hadoop.hive.metastore.api.Table hms
...
+    } else if (!schemaDifference.getTypeChanged().isEmpty()) {
+      FieldSchema updatedField = schemaDifference.getTypeChanged().get(0);
+      updateSchema.updateColumn(updatedField.getName(), getPrimitiveTypeOrThrow(updatedField),
+          updatedField.getComment());
+
+    // case 3: only comment changed
+    } else if (!schemaDifference.getCommentChanged().isEmpty()) {
+      FieldSchema updatedField = schemaDifference.getCommentChanged().get(0);
+      updateSchema.updateColumnDoc(updatedField.getName(), updatedField.getComment());
+    }
```
Comment (szlta) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:571]: Will this handle the case when name is unchanged, but both the type and the comment has been altered?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 4]
```diff
@@ -86,6 +88,10 @@
   private static final Splitter TABLE_NAME_SPLITTER = Splitter.on("..");
   private static final String TABLE_NAME_SEPARATOR = "..";
 
+  private static final List<AlterTableType> ALLOWED_ALTER_OPS = ImmutableList.of(
+      AlterTableType.ADDPROPS, AlterTableType.DROPPROPS, AlterTableType.ADDCOLS,
+      AlterTableType.REPLACE_COLUMNS, AlterTableType.RENAME_COLUMN, AlterTableType.SETPARTITIONSPEC);
+
```
[the remark this answers is not in the document]
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:94]: Yes, good idea!

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 3 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:571]: Yes, it should. In that case, we'd have an entry in both the `commentChanged` and the `typeChanged` lists in the `schemaDifference`. There's a unit test covering this called `testAlterTableChangeColumnTypeAndComment`

Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:568]: Thanks!

**Review by szlta — APPROVED** [review]

**Pull request merged**